### PR TITLE
Consider member-level nonwritable decorations

### LINKED ISSUE #73: Fix NonWritable check when vertexPipelineStoresAndAtomics not enabled
**Issue by [greg-lunarg](https://github.com/greg-lunarg) (MIGRATED)**
_Wednesday Mar 28, 2018 at 21:29 GMT_
_Originally opened as https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/2526_

----

The following Vulkan validation layer error starting popping up on code that previously executed without error:

SC(ERROR / SPEC): object: 0x0 type: 0 msgNum: 15 - Shader requires VkPhysicalDeviceFeatures::vertexPipelineStoresAndAtom
ics but is not enabled on the device
SC(ERROR): object: 0x0 type: 0 location: 1167 msg_code: 15: Object: 0x0 | Shader requires VkPhysicalDeviceFeatures::vert
exPipelineStoresAndAtomics but is not enabled on the device

We believe that this error is being reported due to the following recent change to Vulkan validation layers:

KhronosGroup/Vulkan-LoaderAndValidationLayers#2466

From the PR: While walking the shader for used descriptors, also determine whether any used buffers/images/texelbuffers are not decorated with NonWritable.

From the Vulkan spec:

vertexPipelineStoresAndAtomics indicates whether storage buffers and images support stores and atomic operations in the vertex, tessellation, and geometry shader stages. If this feature is not enabled, all storage image, storage texel buffers, and storage buffer variables used by these stages in shader modules must be decorated with the NonWriteable decoration (or the readonly memory qualifier in GLSL).

Here is the shader:

#version 310 es

layout(location = 0) in vec3 in_pos;
layout(location = 1) in vec3 in_normal;

layout(std140, set = 0, binding = 0) readonly buffer param_block {
vec3 light_pos;
vec3 light_color;
mat4 model;
mat4 view_projection;
float alpha;
} params;

layout(location = 0) out vec3 color;
layout(location = 1) out float alpha;

void main()
{
vec3 world_light = vec3(params.model * vec4(params.light_pos, 1.0));
vec3 world_pos = vec3(params.model * vec4(in_pos, 1.0));
vec3 world_normal = mat3(params.model) * in_normal;

vec3 light_dir = world_light - world_pos;
float brightness = dot(light_dir, world_normal) / length(light_dir) / length(world_normal);
brightness = abs(brightness);

gl_Position = params.view_projection * vec4(world_pos, 1.0);
color = params.light_color * brightness;
alpha = params.alpha;
}

and here is the SPIR-V:

// Module Version 10000
// Generated by (magic number): 80005
// Id's are bound by 114

                          Capability Shader
           1:             ExtInstImport  "GLSL.std.450"
                          MemoryModel Logical GLSL450
                          EntryPoint Vertex 4  "main" 38 67 89 102 109
                          Source ESSL 310
                          Name 4  "main"
                          Name 9  "world_light"
                          Name 12  "param_block"
                          MemberName 12(param_block) 0  "light_pos"
                          MemberName 12(param_block) 1  "light_color"
                          MemberName 12(param_block) 2  "model"
                          MemberName 12(param_block) 3  "view_projection"
                          MemberName 12(param_block) 4  "alpha"
                          Name 14  "params"
                          Name 34  "world_pos"
                          Name 38  "in_pos"
                          Name 49  "world_normal"
                          Name 67  "in_normal"
                          Name 70  "light_dir"
                          Name 75  "brightness"
                          Name 87  "gl_PerVertex"
                          MemberName 87(gl_PerVertex) 0  "gl_Position"
                          MemberName 87(gl_PerVertex) 1  "gl_PointSize"
                          Name 89  ""
                          Name 102  "color"
                          Name 109  "alpha"
                          MemberDecorate 12(param_block) 0 NonWritable
                          MemberDecorate 12(param_block) 0 Offset 0
                          MemberDecorate 12(param_block) 1 NonWritable
                          MemberDecorate 12(param_block) 1 Offset 16
                          MemberDecorate 12(param_block) 2 ColMajor
                          MemberDecorate 12(param_block) 2 NonWritable
                          MemberDecorate 12(param_block) 2 Offset 32
                          MemberDecorate 12(param_block) 2 MatrixStride 16
                          MemberDecorate 12(param_block) 3 ColMajor
                          MemberDecorate 12(param_block) 3 NonWritable
                          MemberDecorate 12(param_block) 3 Offset 96
                          MemberDecorate 12(param_block) 3 MatrixStride 16
                          MemberDecorate 12(param_block) 4 NonWritable
                          MemberDecorate 12(param_block) 4 Offset 160
                          Decorate 12(param_block) BufferBlock
                          Decorate 14(params) DescriptorSet 0
                          Decorate 14(params) Binding 0
                          Decorate 38(in_pos) Location 0
                          Decorate 67(in_normal) Location 1
                          MemberDecorate 87(gl_PerVertex) 0 BuiltIn Position
                          MemberDecorate 87(gl_PerVertex) 1 BuiltIn PointSize
                          Decorate 87(gl_PerVertex) Block
                          Decorate 102(color) Location 0
                          Decorate 109(alpha) Location 1
           2:             TypeVoid
           3:             TypeFunction 2
           6:             TypeFloat 32
           7:             TypeVector 6(float) 3
           8:             TypePointer Function 7(fvec3)
          10:             TypeVector 6(float) 4
          11:             TypeMatrix 10(fvec4) 4
12(param_block): TypeStruct 7(fvec3) 7(fvec3) 11 11 6(float)
13: TypePointer Uniform 12(param_block)
14(params): 13(ptr) Variable Uniform
15: TypeInt 32 1
16: 15(int) Constant 2
17: TypePointer Uniform 11
20: 15(int) Constant 0
21: TypePointer Uniform 7(fvec3)
24: 6(float) Constant 1065353216
37: TypePointer Input 7(fvec3)
38(in_pos): 37(ptr) Variable Input
52: TypeMatrix 7(fvec3) 3
53: 6(float) Constant 0
67(in_normal): 37(ptr) Variable Input
74: TypePointer Function 6(float)
87(gl_PerVertex): TypeStruct 10(fvec4) 6(float)
88: TypePointer Output 87(gl_PerVertex)
89: 88(ptr) Variable Output
90: 15(int) Constant 3
99: TypePointer Output 10(fvec4)
101: TypePointer Output 7(fvec3)
102(color): 101(ptr) Variable Output
103: 15(int) Constant 1
108: TypePointer Output 6(float)
109(alpha): 108(ptr) Variable Output
110: 15(int) Constant 4
111: TypePointer Uniform 6(float)

Each of the buffer variables in this shader (members of the shader storage block param_block) is decorated NonWritable, so no error should be emitted.

### PR DESCRIPTION
**Pull request by chrisforbes**
Previously we only cared about whether the NonWritable decoration was
applied to an entire descriptor-backed OpVariable. This is problematic
for real-world shaders, because glslang emits per-member decorations
instead in all cases.

Adjust the logic to consider per-member decorations in these cases.
Fixes #73.

### CODE REVIEW

**Comment by jzulauf-lunarg** [comment]
@zeux  - pull request is intended to address your issue.  If you had the time to test vs. your use case this would be greatly appreciated.

File: layers/shader_validation.cpp [hunk 1]
```diff
@@ -688,37 +688,45 @@ static std::vector<std::pair<uint32_t, interface_var>> CollectInterfaceByInputAt
     return out;
 }
 
-static bool IsWritableDescriptorType(shader_module const *module, uint32_t type_id) {
+static bool IsWritableDescriptorType(shader_module const *module, uint32_t type_id, bool is_storage_buffer,
```
Comment (jzulauf-lunarg) [inline layers/shader_validation.cpp]: If the entire object is non-writable, can we short circuit the bulk of this function for the false case?

File: layers/shader_validation.cpp [hunk 2]
```diff
@@ -688,37 +688,45 @@ static std::vector<std::pair<uint32_t, interface_var>> CollectInterfaceByInputAt
...
     switch (type.opcode()) {
         case spv::OpTypeImage: {
             auto dim = type.word(3);
             auto sampled = type.word(7);
-            return sampled == 2 && dim != spv::DimSubpassData;
+            return sampled == 2 && dim != spv::DimSubpassData && !entire_object_nonwritable;
         }
 
-        case spv::OpTypeStruct:
+        case spv::OpTypeStruct: {
```
Comment (jzulauf-lunarg) [inline layers/shader_validation.cpp]: Do we need to recur into struct member types for full coverage, or is this code guarantee to be at a leaf type?  For example if a contained struct was decorated as non-writable, but all of *it's* members were...

Comment (chrisforbes) [reply layers/shader_validation.cpp]: According to my reading of the spec, this only applies to direct members of block types;  The struct we're encountering here is the block itself.

GLSL disallows memory qualifiers on members of normal structs, even if that struct is then used in a block declaration.

File: layers/shader_validation.cpp (hunk 1 above)
Comment (chrisforbes) [reply layers/shader_validation.cpp]: We could today, yes.

I'm not sure we should long-term though -- this code will fairly naturally extend to support validating the NonReadable decoration as well.

Comment (jzulauf-lunarg) [reply layers/shader_validation.cpp]: then leave it.

File: layers/shader_validation.cpp [hunk 3]
```diff
@@ -688,37 +688,45 @@ static std::vector<std::pair<uint32_t, interface_var>> CollectInterfaceByInputAt
...
+                        // has its struct type decorated with BufferBlock.
+                        is_storage_buffer = true;
                     }
+                } else if (insn.opcode() == spv::OpMemberDecorate && insn.word(1) == type.word(1) &&
+                           insn.word(3) == spv::DecorationNonWritable) {
+                    nonwritable_members.insert(insn.word(2));
                 }
             }
+
+            // A buffer is writable if it's either flavor of storage buffer, and has any member not decorateddoesn't have all of its
```
Comment (tobine) [inline layers/shader_validation.cpp]: Looks like an incomplete comment update here. Also did you clang-format as this looks a bit long.

File: layers/shader_validation.cpp [hunk 4]
```diff
@@ -766,7 +776,8 @@ static std::vector<std::pair<descriptor_slot_t, interface_var>> CollectInterface
             v.type_id = insn.word(1);
             out.emplace_back(std::make_pair(set, binding), v);
 
-            if (var_nonwritable.find(id) == var_nonwritable.end() && IsWritableDescriptorType(src, insn.word(1))) {
+            if (IsWritableDescriptorType(src, insn.word(1), insn.word(3) == spv::StorageClassStorageBuffer,
+                                         var_nonwritable.find(id) != var_nonwritable.end())) {
```
Comment (tobine) [inline layers/shader_validation.cpp]: If we know at this point that the entire object is non-writable can't we skip the call down to IsWriteableDescriptorType? I suspect there's a reason that you refactored to pass this info to the function but it's not clear to me.

File: layers/shader_validation.cpp (hunk 3 above)
Comment (chrisforbes) [reply layers/shader_validation.cpp]: Passes clang-format. Something weird happened here, though...

File: layers/shader_validation.cpp (hunk 4 above)
Comment (chrisforbes) [reply layers/shader_validation.cpp]: I'm going to undo this. It made a lot more sense in some of the early versions of the change.

**Comment by chrisforbes** [comment]
Updated.

**Comment by chrisforbes** [comment]
I had some ideas while writing this about the direction to go for per-descriptor NonWritable and NonReadable tracking which we'll need "sometime", but adjusted this patch to just do exactly what it needs to for now.

**Pull request merged**